server/src/instant/db/transaction.clj

```diff
@@ -141,36 +146,119 @@
+  [conn app-id entity-ids]
+  (if (empty? entity-ids)
+    {}
+    (->>  {:from            :triples
+           :join            [:attrs [:= :triples.attr_id :attrs.id]
+                             :idents [:= :attrs.forward_ident :idents.id]]
+           :where           [:and
+                             [:= :triples.app_id app-id]
+                             [:in :triples.entity_id entity-ids]]
+           :select-distinct [:triples.entity_id :idents.etype]}
+          hsql/format
+          (sql/do-execute! conn)
+          (coll/group-by-to :triples/entity_id :idents/etype))))
+
+(comment
+  (resolve-etypes
+   (instant.jdbc.aurora/conn-pool)
+   #uuid "0e563ace-d25f-44f6-ae00-4e6b9b6d1b2e"
+   [#uuid "4d39508b-9ee2-48a3-b70d-8192d9c5a059"
+    #uuid "005a8767-c0e7-4158-bb9a-62ce1a5858ed"
+    #uuid "005b08a1-4046-4fba-b1d1-a78b0628901c"]))
+
+
+(defn resolve-lookups-for-delete-entity [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)
+        lookup-refs                     (->> delete-entity-steps
+                                             (map second)
+                                             (filter triple-model/eid-lookup-ref?))
+        resolved                        (resolve-lookups conn app-id lookup-refs)
+        delete-entity-steps'            (for [[op eid etype] delete-entity-steps]
+                                          [op (get resolved eid eid) etype])]
+    (concat delete-entity-steps' tx-steps')))
+
+(defn resolve-etypes-for-delete-entity [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)
+        untyped-ids                     (->> delete-entity-steps
+                                             (keep (fn [[_ id etype]]
+                                                     (when (nil? etype)
+                                                       id))))
+        resolved                        (resolve-etypes conn app-id untyped-ids)
+        delete-entity-steps'            (for [[op eid etype] delete-entity-steps
+                                              etype'         (if etype
+                                                               [etype]
+                                                               (get resolved eid))]
+                                          [op eid etype'])]
+    (concat delete-entity-steps' tx-steps')))
+
+(defn expand-delete-entity-cascade [conn app-id tx-steps]
+  (let [[delete-entity-steps tx-steps'] (coll/split-by #(= :delete-entity (first %)) tx-steps)]
+    (if (empty? delete-entity-steps)
+      tx-steps
+      (let [ids+etypes           (map next delete-entity-steps)
+            query                (str "WITH RECURSIVE entids (entity_id, etype) AS (
+                                       -- start with entid
+                                         VALUES " (string/join ", " (repeat (count ids+etypes) "(?, ?)")) "
```

Comment:
I would use honeysql here to keep things consistent. I know you're only inserting question marks, but seeing `str` around sql queries makes me a little nervous.